### PR TITLE
kernel snap and gadget snap

### PR DESCRIPTION
**Pull request by walimis**
As we have discussed, I use snapcraft to implement kernel snap and gadget snap.
For 14.04, I have create another layer to build snapcraft on 14.04:
https://github.com/walimis/meta-snappy-tools
An example layer to use this kernel snap and gadget snap is here:
https://github.com/walimis/meta-artik-snappy

### CODE REVIEW

**Comment by morphis** [comment]
Sorry, I didn't had time to look into this yet but will do early next week.

File: classes/gadgetsnap.bbclass [hunk 1]
```diff
@@ -0,0 +1,9 @@
+inherit python3native
+
+#generate gadget snap
```
Comment (morphis) [inline classes/gadgetsnap.bbclass]: Comment is not needed here.

File: classes/gadgetsnap.bbclass [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+do_gadget_snap() {
+    snapcraft snap ${WORKDIR}/gadget --output gadget.snap
```
Comment (morphis) [inline classes/gadgetsnap.bbclass]: Why do we use snapcraft and not just mksquashfs?

File: classes/gadgetsnap.bbclass [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+}
+
+addtask gadget_snap after do_patch before do_compile
+
```
Comment (morphis) [inline classes/gadgetsnap.bbclass]: Unneeded empty trailing line.

File: classes/kernelsnap.bbclass [hunk 4]
```diff
@@ -0,0 +1,11 @@
+inherit python3native
+
+#generate kernel snap
```
Comment (morphis) [inline classes/kernelsnap.bbclass]: Comment is not needed here.

File: classes/kernelsnap.bbclass [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+do_kernel_snap () {
+    set +e
+    cd ${S}
+    snapcraft --target-arch armhf snap --output kernel.snap
```
Comment (morphis) [inline classes/kernelsnap.bbclass]: See comment above.

File: classes/kernelsnap.bbclass [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+}
+
+addtask kernel_snap after do_patch before do_compile
+
```
Comment (morphis) [inline classes/kernelsnap.bbclass]: Unneeded empty trailing line.

File: recipes-bsp/gadget-snap/gadget-snap.inc [hunk 7]
```diff
@@ -0,0 +1,11 @@
+inherit gadgetsnap
```
Comment (morphis) [inline recipes-bsp/gadget-snap/gadget-snap.inc]: Name bbclass gadget_snap

File: recipes-bsp/gadget-snap/gadget-snap.inc [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+do_install() {
+    install -d  ${D}/
+    install -m 0755 ${S}/gadget.snap ${D}/
+}
+
+FILES_${PN} = " \
+    /gadget.snap \
+"
+
```
Comment (morphis) [inline recipes-bsp/gadget-snap/gadget-snap.inc]: Unneeded empty trailing line.

File: recipes-kernel/linux/kernel-snap.inc [hunk 9]
```diff
@@ -0,0 +1,11 @@
+inherit kernelsnap
```
Comment (morphis) [inline recipes-kernel/linux/kernel-snap.inc]: Name the bbclass kernel_snap

File: recipes-kernel/linux/kernel-snap.inc [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+FILES_kernel-base += " \
+    /kernel.snap \
+    "
+
+do_install_append () {
+    install -m 0666 ${S}/kernel.snap ${D}/
+}
+
+
```
Comment (morphis) [inline recipes-kernel/linux/kernel-snap.inc]: Unneeded empty trailing line.

File: recipes-kernel/linux/kernel-snap.inc [hunk 11]
```diff
@@ -0,0 +1,11 @@
+inherit kernelsnap
+
+FILES_kernel-base += " \
+    /kernel.snap \
```
Comment (morphis) [inline recipes-kernel/linux/kernel-snap.inc]: use ${D}/kernel.snap

File: recipes-kernel/linux/kernel-snap.inc [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+    "
+
+do_install_append () {
+    install -m 0666 ${S}/kernel.snap ${D}/
```
Comment (morphis) [inline recipes-kernel/linux/kernel-snap.inc]: Use just ${D}

File: recipes-kernel/linux/linux-yocto-snappy_3.10.inc [hunk 13]
```diff
@@ -0,0 +1,35 @@
+FILESEXTRAPATHS_prepend := "${THISDIR}/files:"
+
+SRC_URI += " \
```
Comment (morphis) [inline recipes-kernel/linux/linux-yocto-snappy_3.10.inc]: Please add a comment explain what all these patches are for.

File: recipes-bsp/gadget-snap/gadget-snap.inc (hunk 7 above)
Comment (morphis) [inline recipes-bsp/gadget-snap/gadget-snap.inc]: Also this should be merged into a single bbclass not an .inc file.

File: recipes-kernel/linux/linux-yocto-snappy_3.10.inc [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	file://0023-kbuild-mergeconfig-remove-redundant-objtree.patch \
+	file://0024-kbuild-add-generic-mergeconfig-target-.config.patch \
+	file://0025-UBUNTU-Config-snappy-config-fragments-generic-system.patch \
+	file://0026-UBUNTU-Config-snappy-snappy-SQUASHFS-m.patch \
+	file://0027-UBUNTU-SAUCE-add-a-snapcraft.yaml-to-create-a-snappy.patch \
+	file://0028-UBUNTU-SAUCE-add-a-snapcraft.yaml-to-create-a-snappy.patch \
+	file://0029-UBUNTU-Config-snappy-containers-lxc.config-was-super.patch \
+	file://0030-UBUNTU-Config-snappy-containers-CONFIG_OVERLAY_FS_V1.patch \
+    "
+include kernel-snap.inc
```
Comment (morphis) [inline recipes-kernel/linux/linux-yocto-snappy_3.10.inc]: This should be a kernel_snap.bbclass rather than an .inc file.

**Comment by morphis** [comment]
Also can we merge meta-snappy-tools into this layer too? I don't see any reason to have multiple layers. A single meta-snappy should be enough.

**Comment by bboozzoo** [comment]
The repo has moved to https://github.com/snapcore/meta-snapd. Please open a PR there.

**Pull request closed without merging**